packages/flutter_web_plugins/lib/src/history/url_strategy.dart

```diff
@@ -0,0 +1,332 @@
+
+  /// Moves forwards or backwards through the history stack.
+  Future<void> go(int count);
+}
+
+/// This is an implementation of [UrlStrategy] that uses the browser URL's
+/// [hash fragments](https://en.wikipedia.org/wiki/Uniform_Resource_Locator#Syntax)
+/// to represent its state.
+///
+/// In order to use this [UrlStrategy] for an app, it needs to be set like this:
+///
+/// ```dart
+/// import 'package:flutter_web_plugins/flutter_web_plugins.dart';
+///
+/// // Somewhere before calling `runApp()` do:
+/// setUrlStrategy(const HashUrlStrategy());
+/// ```
+class HashUrlStrategy extends UrlStrategy {
+  /// Creates an instance of [HashUrlStrategy].
+  ///
+  /// The [PlatformLocation] parameter is useful for testing to avoid
+  /// interacting with the actual browser.
+  const HashUrlStrategy(
+      [this._platformLocation = const BrowserPlatformLocation()]);
+
+  final PlatformLocation _platformLocation;
+
+  @override
+  ui.VoidCallback onPopState(html.EventListener fn) {
+    _platformLocation.onPopState(fn);
+    return () => _platformLocation.offPopState(fn);
+  }
+
+  @override
+  String getPath() {
+    // the hash value is always prefixed with a `#`
+    // and if it is empty then it will stay empty
+    final String path = _platformLocation.hash ?? '';
+    assert(path.isEmpty || path.startsWith('#'));
+
+    // We don't want to return an empty string as a path. Instead we default to "/".
+    if (path.isEmpty || path == '#') {
+      return '/';
+    }
+    // At this point, we know [path] starts with "#" and isn't empty.
+    return path.substring(1);
+  }
+
+  @override
+  dynamic getState() => _platformLocation.state;
+
+  @override
+  String prepareExternalUrl(String internalUrl) {
+    // It's convention that if the hash path is empty, we omit the `#`; however,
+    // if the empty URL is pushed it won't replace any existing fragment. So
+    // when the hash path is empty, we instead return the location's path and
+    // query.
+    return internalUrl.isEmpty
+        ? '${_platformLocation.pathname}${_platformLocation.search}'
+        : '#$internalUrl';
+  }
+
+  @override
+  void pushState(dynamic state, String title, String url) {
+    _platformLocation.pushState(state, title, prepareExternalUrl(url));
+  }
+
+  @override
+  void replaceState(dynamic state, String title, String url) {
+    _platformLocation.replaceState(state, title, prepareExternalUrl(url));
+  }
+
+  @override
+  Future<void> go(int count) {
+    _platformLocation.go(count);
+    return _waitForPopState();
+  }
+
+  /// Waits until the next popstate event is fired.
+  ///
+  /// This is useful for example to wait until the browser has handled the
+  /// `history.back` transition.
+  Future<void> _waitForPopState() {
+    final Completer<void> completer = Completer<void>();
+    ui.VoidCallback unsubscribe;
+    unsubscribe = onPopState((_) {
+      unsubscribe();
+      completer.complete();
+    });
```

Comment:
I wanted this to work, but it didn't :(

I think the problem may be that the analyzer sees the reference to `unsubscribe` in the RHS expression and doesn't allow it because the declaration of `unsubscribe` hasn't happened yet. The analyzer doesn't know that the callback will be called *after* the declaration is done. Theoretically, the callback could be called immediately by `onPopState` and `unsubscribe` wouldn't have been declared by then. `late` helps with late *assignment* but not late *declaration*.

What you are suggesting is equivalent to:
```dart
var $tmp123 = onPopState((_) {
  unsubscribe();
  completer.complete();
});
late ui.VoidCallback unsubscribe = $tmp123;
```